src/interlocked_hl.c

```diff
@@ -163,6 +163,53 @@ IMPLEMENT_MOCKABLE_FUNCTION(, INTERLOCKED_HL_RESULT, InterlockedHL_WaitForNotVal
     return result;
 }
 
+IMPLEMENT_MOCKABLE_FUNCTION(, INTERLOCKED_HL_RESULT, InterlockedHL_CompareExchangeIf, int32_t volatile_atomic*, target, int32_t, exchange, INTERLOCKED_COMPARE_EXCHANGE_IF, compare, int32_t*, original_target)
+{
+    INTERLOCKED_HL_RESULT result;
+    if (
+        /* Codes_SRS_INTERLOCKED_HL_01_009: [ If target is NULL then InterlockedHL_CompareExchangeIf shall return fail and return INTERLOCKED_HL_ERROR. ]*/
+        (target == NULL) ||
+        /* Codes_SRS_INTERLOCKED_HL_01_010: [ If compare is NULL then InterlockedHL_CompareExchangeIf shall return fail and return INTERLOCKED_HL_ERROR. ]*/
+        (compare == NULL) ||
+        /* Codes_SRS_INTERLOCKED_HL_01_011: [ If original_target is NULL then InterlockedHL_CompareExchangeIf shall return fail and return INTERLOCKED_HL_ERROR. ]*/
+        (original_target == NULL)
+        )
+    {
+        LogError("invalid arguments int64_t volatile_atomic* target=%p, int32_t exchange=%" PRId32 ", INTERLOCKED_COMPARE_EXCHANGE_IF compare=%p original_target=%p",
+            target, exchange, compare, original_target);
+        result = INTERLOCKED_HL_ERROR;
+    }
+    else
+    {
+        /* Codes_SRS_INTERLOCKED_HL_01_012: [ InterlockedHL_CompareExchangeIf shall acquire the initial value of target. ]*/
+        int32_t copyOfTarget = interlocked_add(target, 0);
+
+        /* Codes_SRS_INTERLOCKED_HL_01_013: [ If compare(target, exchange) returns true then InterlockedHL_CompareExchangeIf shall exchange target with exchange. ]*/
+        if (compare(copyOfTarget, exchange))
+        {
+            if (interlocked_compare_exchange(target, exchange, copyOfTarget) == copyOfTarget)
+            {
+                /* Codes_SRS_INTERLOCKED_HL_01_015: [ If target did not change meanwhile then InterlockedHL_CompareExchangeIf shall return INTERLOCKED_HL_OK and shall peform the exchange of values. ]*/
+                result = INTERLOCKED_HL_OK;
+            }
+            else
+            {
+                /* Codes_SRS_INTERLOCKED_HL_01_014: [ If target changed meanwhile then InterlockedHL_CompareExchangeIf shall return INTERLOCKED_HL_CHANGED and shall not peform any exchange of values. ]*/
+                result = INTERLOCKED_HL_CHANGED;
+            }
+        }
+        else
+        {
+            /* Codes_SRS_INTERLOCKED_HL_01_017: [ If compare returns false then  InterlockedHL_CompareExchangeIf shall not perform any exchanges and return INTERLOCKED_HL_OK. ]*/
+            result = INTERLOCKED_HL_OK;
```

Comment:
>result = INTERLOCKED_HL_OK; [](http://example.com/codeflow?start=12&length=27)

One thing that makes me think, should we change the return value here to differentiate between a positive compare and a negative compare?

Reply:
I hear you. Let's leave that for when we hit that case though.